cloudinit/util.py

```diff
@@ -79,12 +79,20 @@ def wrapper(f):
 
 
 @lru_cache()
-def get_architecture(target=None):
+def _get_architecture(target=None):
     out, _ = subp(['dpkg', '--print-architecture'], capture=True,
                   target=target)
     return out.strip()
 
 
+def get_architecture(target=None):
+    if target_path(target) != "/":
+        # do not use or update cache if target is provided
```

Comment:
i'm not sure this comment isn't just a lie

Reply:
Looks like when the @lru_cache() was added in 2a135c4a  the comment was already present in the _lsb_release implementation and was retained.  So yes, let's clarify that and verify that that get_architecture(target=None)  != get_architecture(target=/path/to/chroot/somewhere).

IIUC the python cache will capture args,kwargs as part of the hash when caching, so calling with different params should not return the same cached value.

```
#!/usr/bin/python3
from functools import lru_cache
import os
import tempfile
import shutil


def touch(fname, times=None):
    fhandle = open(fname, 'a')
    try:
        os.utime(fname, times)
    finally:
        fhandle.close()


def listdir_fullpath(d):
    return [os.path.join(d, f) for f in os.listdir(d)]


def listdir(target):
    return listdir_fullpath(target)


@lru_cache()
def cached_listdir(target):
    return listdir_fullpath(target)


dir1 = dir2 = None
try:
    dir1 = tempfile.mkdtemp(suffix='-dir1')
    touch(os.path.join(dir1, 'file1'))
    print('Uncached listdir(%s): %s' % (dir1, listdir(dir1)))
    print('  Cached listdir(%s): %s' % (dir1, cached_listdir(dir1)))
    print('Creating new file in %s' % dir1)
    touch(os.path.join(dir1, 'file2'))
    print('Uncached listdir(%s): %s' % (dir1, listdir(dir1)))
    print('  Cached listdir(%s): %s' % (dir1, cached_listdir(dir1)))
    print('Creating new directory with file3')
    dir2 = tempfile.mkdtemp(suffix='-dir2')
    touch(os.path.join(dir2, 'file3'))
    print('Uncached listdir(%s): %s' % (dir2, listdir(dir2)))
    print('  Cached listdir(%s): %s' % (dir2, cached_listdir(dir2)))

finally:
    for d in [dir1, dir2]:
        if d and os.path.exists(d):
            shutil.rmtree(d)


EXAMPLE_OUTPUT="""
Uncached listdir(/tmp/tmp_pq5szl7-dir1): ['/tmp/tmp_pq5szl7-dir1/file1']
  Cached listdir(/tmp/tmp_pq5szl7-dir1): ['/tmp/tmp_pq5szl7-dir1/file1']
Creating new file in /tmp/tmp_pq5szl7-dir1
Uncached listdir(/tmp/tmp_pq5szl7-dir1): ['/tmp/tmp_pq5szl7-dir1/file1', '/tmp/tmp_pq5szl7-dir1/file2']
  Cached listdir(/tmp/tmp_pq5szl7-dir1): ['/tmp/tmp_pq5szl7-dir1/file1']
Creating new directory with file3
Uncached listdir(/tmp/tmprx227pa7-dir2): ['/tmp/tmprx227pa7-dir2/file3']
  Cached listdir(/tmp/tmprx227pa7-dir2): ['/tmp/tmprx227pa7-dir2/file3']
"""
```

Reply:
https://stackoverflow.com/questions/49883177/how-does-lru-cache-from-functools-work